api/envoy/api/v2/lds.proto

```diff
@@ -203,4 +217,13 @@ message Listener {
   // <envoy_api_field_listener.UdpListenerConfig.udp_listener_name>` = "raw_udp_listener" for
   // creating a packet-oriented UDP listener. If not present, treat it as "raw_udp_listener".
   listener.UdpListenerConfig udp_listener_config = 18;
+
+  // [#not-implemented-hide:]
+  // Used only when the :ref:`type<envoy_api_field_Listener.type>` field is set to
+  // :ref:`API_HTTP<envoy_api_field_Listener.Type.API_HTTP>`.
+  // Maps port to corresponding HttpConnectionManager configuration.
+  // When an HTTP client is trying to connect to "servername:port", it will look up the port in
+  // this map to find the root of its configuration.
+  map<uint32, envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager>
```

Comment:
I think this is a bit of a strange interface TBH. What does port mean in the world of an API HTTP listener?

What's strange is we're thinking here about a tuple of destination server host name and port, and we're only providing a binding on the port. In general, we should either have the capability here to bind to both server host name (analogous to SNI) and port, and add a general bind configuration for this for API HTTP listeners, or we should defer both server name and port to the `RouteConfiguration` as L7 concerns.

Also, I don't think you need a map here. Each listener should only be concerned with a single port, just as we do today for regular TCP. The xDS APIs already have a notion of multiple port listenening, you have multiples `Listener` objects; gRPC should process multiple listeners to deal with this.

Reply:
> What's strange is we're thinking here about a tuple of destination server host name and port, and we're only providing a binding on the port. In general, we should either have the capability here to bind to both server host name (analogous to SNI) and port, and add a general bind configuration for this for API HTTP listeners, or we should defer both server name and port to the `RouteConfiguration` as L7 concerns.
> Also, I don't think you need a map here. Each listener should only be concerned with a single port, just as we do today for regular TCP. The xDS APIs already have a notion of multiple port listenening, you have multiples Listener objects; gRPC should process multiple listeners to deal with this.

+1, we also have FilterChain (+Match) to do bind host name, port, address, etc. So the mapping from port to HCM can be achieved with them too. We just need a better capabilities to say this xDS client only support HCM but not other L4 filters in the FilterChain. This way the same mechanism can be applied to other protocols with different capabilities.

Reply:
This is not quite what I had in mind, but close. Related to @htuch comment, IMO it would be better here to avoid server/port since it's confusing. What I had in mind was just an opaque name that is assigned to an API listener. As such, we would not need a map here, but just a name of the API listener, which should be unique (we might actually just be able to use `name`). Then, it's an implementation detail how `name` is actually exposed in the library code.

cc @junr03 @goaway please read and comment on ^ as this is highly relevant to Envoy Mobile.

Reply:
Thanks for all the comments -- I expected this part to generate discussion, and these are exactly the types of issues that I wanted to surface.

My first thought about how to handle ports here was actually to have a way to attach a port number to a `VirtualHost` object, in which case it would be taken into account as part of choosing the `VirtualHost` for each request.  However, this seems drastically different from how Envoy (or any other proxy using xDS) would expect to handle destination ports, and we were concerned that this difference would make it harder for existing management servers (and their underlying infrastructures) to support gRPC.

AIUI, in Envoy, ports are always handled in the `Listener`, either by having a different `Listener` for each port or by having iptables forward all traffic to a single `Listener` that contains a different filter chain for each destination port.  Because there can be only one `RouteConfiguration` per filter chain, and the list of `VirtualHost`s is defined within the `RouteConfiguration`, the port is effectively an implicit part of the scope in which the `VirtualHost` is chosen, so there's no need to use the port when choosing the `VirtualHost`.

The approach I proposed here was an attempt to more closely match the data model currently used by proxies by having the port handled in the `Listener`.  Each entry in the map is conceptually a different filter chain, except that each filter chain supports only one match criterion (the port) and one filter (HCM).  I haven't fully thought through how to handle the case where there is actually a separate listener for each port, but perhaps we could handle that by defining multiple HTTP API listeners, differentiated by name.  (This is somewhat related to the discussion above about how the LDS request will look from gRPC clients.)

I do agree that it seems confusing to have the port matching done here and the server name matching done in the `VirtualHost`.  But this is exactly the way that Envoy does things today.  An alternative might be to somehow move port-matching into the `VirtualHost` layer even for Envoy, so that we could not deal with this in the HTTP API listener but still be consistent with how this works in Envoy.  But it's not obvious to me if anyone would actually use that if they already have a different listener filter chain for each port anyway.  (I am guessing that there are cases where it could hurt performance to combine multiple ports into a single filter chain?)

Thoughts...?

Reply:
I think what I'm confused about here is why we need to convolve port at all at this layer for an API listener? Why can't it just be an opaque name and if the client/server want to encode port in that somehow that is fine?

Reply:
I'm not certain I'm fully following here, but it seems to me like having support for multiple listeners with different addressable names (and potentially their own filter chain and/or route configuration) supports everything we'd need. I'm really not sure what having any notion of a (virtual?) port would actually gain us.

Reply:
This is closely related to the thread above about how the LDS request will work -- essentially, these two discussions are two sides of the same issue.

If we decide that the client should be required to request the listener it wants by name in the LDS request, then we can certainly make the port part of that name.  But that approach imposes some restrictions on the management server:

- The xDS server must provide one listener for each virtual host.
- The listeners' names must match the virtual hosts.

I am concerned that there will be cases where these restrictions might be painful.  For example, let's say that you have an existing management server that is used by Envoy.  The management server currently structures its data such that there is a listener per port (or a single listener with each port handled by a different filter chain), it will have a separate `RouteConfiguration` for each port, and each `RouteConfiguration` will have a list of many `VirtualHost`s.

Now let's say that you have some gRPC clients that are currently going through Envoy but that you want to switch to use xDS directly.  If we require the management server to provide a separate listener for each virtual host for gRPC, then the management server's existing data model (having multiple `VirtualHosts` on the same `RouteConfiguration` for a given port) will not work.  It will instead need to be changed to provide a separate `RouteConfiguration` for each virtual host, which seems like a non-trivial change.

Conversely, if we go with the approach shown in this PR, then the existing data model will continue to work, because you can continue to select the `RouteConfiguration` on a per-port basis.  The only change needed on the management server is to add support for the HTTP API listener.

This is why I said that I would prefer to structure this as closely as possible to the way that Envoy deals with things.  I agree that it's a bit counter-intuitive, but I think this is a structural problem in the xDS API, and I think we should either find a way to fix that for Envoy or stick with the current model for gRPC.

Does this make sense?

Reply:
> Does this make sense?

What you are saying makes sense, but I still don't think we should convolve ports, etc. into something that inherently only needs a name. I don't really think it's realistic that someone is going to want to seamlessly switch between Envoy and gRPC or some other client. To me that is a higher level concern.

Reply:
Just FYI, this use-case is not theoretical -- we do actually have customers that have this requirement.

Let's discuss further when we meet on Monday.

Reply:
After some more discussion with folks on our end, we've agreed that we can make this work by having the client query for the listener by name.  So, I've removed the port map and the capability from this PR.  Now all we have in the listener is the HCM config.